### ISSUE TITLE
(My API Change Proposal)

### ISSUE THREAD

**Issue by balt-dev** [rust-lang/libs-team#461]
# Proposal

## Problem statement

`core::cmp::Ordering` is not ergonomic to work with, often requiring explicit match statements and writing out full variant names.

## Motivating examples or use cases

Take, for example, https://docs.rs/num-bigint/0.4.6/src/num_bigint/biguint.rs.html#86.
```rust
#[inline]
fn cmp_slice(a: &[BigDigit], b: &[BigDigit]) -> Ordering {
    match Ord::cmp(&a.len(), &b.len()) {
        Ordering::Equal => Iterator::cmp(a.iter().rev(), b.iter().rev()),
        other => other,
    }
}
```
This requires an explicit match statement, which again requires writing out the type name, OR requires enclosing a function call with `.then` or `.then_with`, both being not very ergonomic.

## Solution sketch

Implementing `Try` seems to be a good fix here, as it would allow quick short-circuiting, avoiding having to write all this out.
See below:

```rust
impl Ord for BigUint {
    #[inline]
    fn cmp(&self, other: &BigUint) -> Ordering {
        cmp_slice(&self.data[..], &other.data[..])
    }
}

#[inline]
fn cmp_slice(a: &[BigDigit], b: &[BigDigit]) -> Ordering {
    a.len().cmp(&b.len()?;
    Iterator::cmp(a.iter().rev(), b.iter().rev())
}
````

## Alternatives

- Do nothing
  Definitely easier, but also leaves it in a slightly awkward spot.

**Comment by balt-dev** [comment]
Forgot to change the title 😅

**Comment by balt-dev** [comment]
I also had the idea of implementing `FromIterator<Ordering>` as an alternative to `Iterator::cmp`, but I came to realize it was probably more verbose than it.
However, I _also_ had the idea of implementing `Neg` as a shorthand for `.reverse()`, which might be just a good change overall.

**Comment by kennytm** [comment]
> This requires an explicit match statement, which again requires writing out the type name, OR requires enclosing a function call with .then or .then_with, both being not very ergonomic.

The equivalent using [`.then_with()`](https://doc.rust-lang.org/std/cmp/enum.Ordering.html#method.then_with) does not look too unergonomic

```rust
fn cmp_slice(a: &[BigDigit], b: &[BigDigit]) -> Ordering {
    Ord::cmp(&a.len(), &b.len())
        .then_with(|| Iterator::cmp(a.iter().rev(), b.iter().rev()))
}
```

"Writing out the type name" is probably just a convention chosen by the author. I don't see any reason why you can't write this as

```rust
fn cmp_slice(a: &[BigDigit], b: &[BigDigit]) -> Ordering {
    a.len().cmp(&b.len())
        .then_with(|| a.iter().rev().cmp(b.iter().rev()))
}
```

**Comment by balt-dev** [comment]
The problem with this is that, say if you have a function that returns a `Result`, you can't `Try` through a closure.

**Comment by balt-dev** [comment]
...which I realize would make this not work anyway. Hm.

**Comment by cuviper** [comment]
Right, you can't `Try` both `Ordering` and `Result` in the same context, unless we also had `FromResidual` between them.
